taichi/program/state_flow_graph.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+#include "taichi/lang_util.h"
+#include "taichi/program/program.h"
+#include "taichi/program/async_utils.h"
+
+TLANG_NAMESPACE_BEGIN
+
+class StateFlowGraph {
+ public:
+  struct Node;
+  using StateToNodeMapping =
+      std::unordered_map<AsyncState, Node *, AsyncStateHash>;
+  // Each node is a task
+  // Note: after SFG is done, each node here should hold a TaskLaunchRecord.
+  // Optimization should happen fully on the SFG, instead of the queue in
+  // AsyncEngine.
+  // Before we migrate, the SFG is only used for visualization. Therefore we
+  // only store a kernel_name, which is used as the label of the GraphViz node.
+  struct Node {
+    //  TODO: make use of IRHandle here
+    IRNode *root;
+    std::string kernel_name;
+    StateToNodeMapping input_edges, output_edges;
```

Comment:
IIUC, for `output_edges`, one `AsyncState` could map to multiple `Node*`? E.g. if both `foo()` and `bar()` needs `x`, then for `init_node`, `(x, value)` will be mapped to two nodes.

Reply:
Let's assume we call `foo()` before `bar()`.

Your statement is true if `foo()` only reads `x` but not write to it.

The **value state** of a field represents the latest value version of a field. For now, even if `foo()` only **partially** modifies `x`, the latest version is held by `foo()`. This helps us to get rid of the phi nodes. Most optimization can still be done.

If `foo()` writes to `x`, then it would be something like

```
          x_value                x_value                   x_value
init  ---------------> foo() ----------------> bar() ---------------> ?
```

Reply:
> Your statement is true if foo() only reads x but not write to it.

Ah yes, that is what i was thinking about..  Say `foo` and `bar` only reads `x`. In this case, do you think it is fine for the node to report **any** node it is transitioning to?

I'm thinking about this from the visualization perspective, i.e. if the output nodes are incomplete, the DAG might be incomplete. OTOH, because input nodes are fine, we should still be able to construct the full DAG..

Reply:
> In this case, do you think it is fine for the node to report **any** node it is transitioning to?

IIUC here we are considering letting the initial node having multiple outputs with edge `x_value`? Yes, that's the desired behavior.



> I'm thinking about this from the visualization perspective, i.e. if the output nodes are incomplete, the DAG might be incomplete. OTOH, because input nodes are fine, we should still be able to construct the full DAG..

If we link all the output nodes, then the output edges and input edges must have a 1-to-1 correspondence. Then we need to visualize the output edges only or input edges only :-)